tidy3d/plugins/waveguide/rectangular_dielectric.py

```diff
@@ -0,0 +1,944 @@
+            lda = wavelength / n
+            return 1.5 * lda.max()
+        return val
+
+    @pydantic.validator("box_thickness", always=True)
+    def _set_box_thickness(cls, val, values):
+        if val is None:
+            wavelength = values["wavelength"]
+            medium = values["box_medium"]
+            n = numpy.array([medium.nk_model(f)[0] for f in C_0 / wavelength])
+            lda = wavelength / n
+            return 1.5 * lda.max()
+        return val
+
+    @pydantic.validator("side_margin", always=True)
+    def _set_side_thickness(cls, val, values):
+        return max(values["clad_thickness"], values["box_thickness"]) if val is None else val
+
+    @pydantic.validator("gap", always=True)
+    def _validate_gaps(cls, val, values):
+        if val.size == 1 and values["core_width"].size != 2:
+            return numpy.array([val[0]] * (values["core_width"].size - 1))
+        if val.size != values["core_width"].size - 1:
+            raise ValidationError("Number of gaps must be 1 less than number of core widths.")
+        return val
+
+    @pydantic.root_validator
+    def _ensure_consistency(cls, values):
+        if values["sidewall_thickness"] > 0 and values["sidewall_medium"] is None:
+            raise ValidationError(
+                "Sidewall medium must be provided when sidewall thickness is greater than 0."
+            )
+
+        if values["sidewall_thickness"] == 0 and values["sidewall_medium"] is not None:
+            log.warning("Sidewall medium not used because sidewall thickness is zero.")
+
+        if values["surface_thickness"] > 0 and values["surface_medium"] is None:
+            raise ValidationError(
+                "Surface medium must be provided when surface thickness is greater than 0."
+            )
+
+        if values["surface_thickness"] == 0 and values["surface_medium"] is not None:
+            log.warning("Surface medium not used because surface thickness is zero.")
+
+        if values["propagation_axis"] == values["normal_axis"]:
+            raise ValidationError("Propagation and normal axes must be different.")
+
+        return values
+
+    @cached_property
+    def lateral_axis(self):
+        """Lateral direction axis"""
+        return 3 - self.propagation_axis - self.normal_axis
+
+    def _swizzle(self, lateral_coord, normal_coord, propagation_coord):
+        """Swap the model coordinates to desired axes"""
+        result = [None, None, None]
+        result[self.lateral_axis] = lateral_coord
+        result[self.propagation_axis] = propagation_coord
+        result[self.normal_axis] = normal_coord
+        return result
+
+    def _translate(self, lateral_coord, normal_coord, propagation_coord):
+        """Swap the model coordinates to desired axes and translate to origin"""
+        result = [
+            a + b
+            for a, b in zip(
+                self.origin,
+                self._swizzle(lateral_coord, normal_coord, propagation_coord),
```

Comment:
for readability, let's assign `self._swizzle(lateral_coord, normal_coord, propagation_coord)` to a variable like `model_coordinate`?